### PR TITLE
Seperate test job by framework

### PR DESCRIPTION
**Pull request by chidozieononiwu**
- Separate test and publish of test results in archetype-sdk-client.yml
- Split test and publish steps into separate template, testing for NetFx on windows and NetCore2.1 on other OS's
- Add PackageDirectory as a property so that test can have more specific scope

### CODE REVIEW

**Comment by chidozieononiwu** [comment]
@Azure/azure-sdk-eng

File: eng/pipelines/templates/steps/test-and-publish.yml [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+      DOTNET_SKIP_FIRST_TIME_EXPERIENCE: 1
+      DOTNET_CLI_TELEMETRY_OPTOUT: 1
+      DOTNET_MULTILEVEL_LOOKUP: 0
+    condition: or( eq(variables['OSName'], 'Windows'), eq( '${{ parameters.Framework }}', 'DotNetCore'))
+  - task: PublishTestResults@2
+    displayName: "Publish Results (${{ parameters.Framework }})"
+    condition: or( eq(variables['OSName'], 'Windows'), eq( '${{ parameters.Framework }}', 'DotNetCore'))
+    inputs:
+      testResultsFiles: "**/${{ parameters.Framework }}*.trx"
+      testRunTitle: "$(OSName) ${{ parameters.Framework }} $(DotNetCoreRuntimeVersion)"
```
Comment (weshaggard) [inline eng/pipelines/templates/steps/test-and-publish.yml]: DotNetCoreRuntimeVersion -> TFM

File: eng/pipelines/templates/steps/test-and-publish.yml [hunk 2]
```diff
@@ -0,0 +1,16 @@
+steps:
+  - script: dotnet test eng/service.proj --filter TestCategory!=Live --framework ${{ parameters.TFM }} --logger "trx;LogFileName=${{ parameters.Framework}}.trx" /p:ServiceDirectory=${{ parameters.ServiceDirectory }} /p:PackageDirectory=${{ parameters.PackageDirectory }}
+    displayName: "Build & Test (${{ parameters.Framework }})"
+    env:
+      DOTNET_SKIP_FIRST_TIME_EXPERIENCE: 1
+      DOTNET_CLI_TELEMETRY_OPTOUT: 1
+      DOTNET_MULTILEVEL_LOOKUP: 0
+    condition: or( eq(variables['OSName'], 'Windows'), eq( '${{ parameters.Framework }}', 'DotNetCore'))
```
Comment (weshaggard) [inline eng/pipelines/templates/steps/test-and-publish.yml]: You should just remove this condition and add a condition on the template for net461 and exclude that if it is non-windows.

Comment (chidozieononiwu) [reply eng/pipelines/templates/steps/test-and-publish.yml]: Conditions are not working on templates.

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml [hunk 3]
```diff
@@ -88,16 +88,13 @@ jobs:
...
-          DOTNET_CLI_TELEMETRY_OPTOUT: 1
-          DOTNET_MULTILEVEL_LOOKUP: 0
-      - task: PublishTestResults@2
-        condition: succeededOrFailed()
-        inputs:
-          testResultsFiles: "**/*.trx"
-          testRunTitle: "$(OSName) DotNet $(DotNetCoreRuntimeVersion)"
-          testResultsFormat: "VSTest"
-          mergeTestResults: true
+      - template: ../steps/test-and-publish.yml
```
Comment (danieljurek) [inline eng/pipelines/templates/jobs/archetype-sdk-client.yml]: Realizing that this comment is coming in a bit late: Why not have `TFM` be a parameter of the matrix entries? The matrix would be more repetitive, but we could avoid the conditionals and having to break out `test-and-publish.yml`

```
        Linux:
          OSName: "Linux"
          OSVmImage: "ubuntu-16.04"
          TFM: netcoreapp2.1
        Windows_NetCoreApp:
          OSName: "Windows"
          OSVmImage: "vs2017-win2016"
          TFM: netcoreapp2.1
        Windows_NetFramework:
          OSName: "Windows"
          OSVmImage: "vs2017-win2016"
          TFM: net461
        MacOs:
          OSName: "MacOS"
          OSVmImage: "macOS-10.13"
          TFM: netcoreapp2.1
```

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml [hunk 4]
```diff
@@ -88,16 +95,16 @@ jobs:
...
-      - script: "dotnet test eng/service.proj --filter TestCategory!=Live --logger trx /p:ServiceDirectory=${{parameters.ServiceDirectory}}"
-        displayName: "Build & Test"
+      - script: dotnet test eng/service.proj --filter TestCategory!=Live --framework $(TFM) --logger "trx;LogFileName=$(TFM).trx" /p:ServiceDirectory=${{ parameters.ServiceDirectory }} /p:IncludeSrc=false
+        displayName: "Build & Test ($(TFM))"
         env:
           DOTNET_SKIP_FIRST_TIME_EXPERIENCE: 1
           DOTNET_CLI_TELEMETRY_OPTOUT: 1
           DOTNET_MULTILEVEL_LOOKUP: 0
       - task: PublishTestResults@2
-        condition: succeededOrFailed()
```
Comment (weshaggard) [inline eng/pipelines/templates/jobs/archetype-sdk-client.yml]: We should keep this condition so that test results still get uploaded if the test command fails.

File: eng/pipelines/templates/steps/test-and-publish.yml [hunk 5]
```diff
@@ -0,0 +1,16 @@
+steps:
+  - script: dotnet test eng/service.proj --filter TestCategory!=Live --framework ${{ parameters.TFM }} --logger "trx;LogFileName=${{ parameters.TFM }}.trx" /p:ServiceDirectory=${{ parameters.ServiceDirectory }} /p:IncludeSrc=false
```
Comment (weshaggard) [inline eng/pipelines/templates/steps/test-and-publish.yml]: I assume you plan to remove this one now?

File: eng/service.proj [hunk 6]
```diff
@@ -2,13 +2,19 @@
...
+  <ItemGroup Condition="'$(PackageDirectory)' == ''">
+    <ExcludeMgmtLib Include="..\sdk\$(ServiceDirectory)\*.Management.*\**\*.csproj;..\sdk\*mgmt*\**\*.csproj;" />
+    <TestProjects Include="..\sdk\$(ServiceDirectory)\*\tests\**\*.csproj" />
+    <SrcProjects Include="..\sdk\$(ServiceDirectory)\**\*.csproj" Exclude="@(TestProjects)" />
+    <ProjectReference Include="@(TestProjects)" Exclude="@(ExcludeMgmtLib)" Condition="'$(IncludeTests)' == 'true'" />
+    <ProjectReference Include="@(SrcProjects)" Exclude="@(ExcludeMgmtLib)" Condition="'$(IncludeSrc)' == 'true'" />
   </ItemGroup>
-  <Import Project="..\sdk\$(ServiceDirectory)\*.projects" />
+
+  <Import Condition="'$(ServiceDirectory)' != ''" Project="..\sdk\$(ServiceDirectory)\*.projects" />
```
Comment (weshaggard) [inline eng/service.proj]: Why did you need to add the condition on this import?

File: sdk/servicebus/ci.yml [hunk 7]
```diff
@@ -20,4 +20,4 @@ pr:
 jobs:
 - template: ../../eng/pipelines/templates/jobs/archetype-sdk-client.yml
```
Comment (weshaggard) [inline sdk/servicebus/ci.yml]: You should undo all the changes to this file.

File: eng/service.proj (hunk 6 above)
Comment (chidozieononiwu) [reply eng/service.proj]: That was from my previous attempt to use the package directory. I have removed that. Thanks for pointing it out.

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml [hunk 8]
```diff
@@ -76,12 +76,19 @@ jobs:
...
           OSVmImage: "vs2017-win2016"
+          TFM: netcoreapp2.1
+        Windows_NetFramework:
+          OSName: "Windows"
+          OSVmImage: "vs2017-win2016"
+          TFM: net461
         MacOs:
           OSName: "MacOS"
           OSVmImage: "macOS-10.13"
+          TFM: netcoreapp2.1
```
Comment (weshaggard) [inline eng/pipelines/templates/jobs/archetype-sdk-client.yml]: Can you please rename TFM to TargetFramework to make it more explicit and also alight with what it is called in the projects?

File: eng/service.proj [hunk 9]
```diff
@@ -2,12 +2,18 @@
...
     <ServiceDirectory Condition="'$(ServiceDirectory)' == '' and '$(Scope)' != ''">$(Scope)</ServiceDirectory>
     <ServiceDirectory Condition="'$(ServiceDirectory)' == ''">*</ServiceDirectory>
+    <IncludeTests Condition="'$(IncludeTests)' == ''">true</IncludeTests>
+    <IncludeSrc Condition="'$(IncludeSrc)' == ''">true</IncludeSrc>
   </PropertyGroup>
 
-  <ItemGroup>
-    <ExcludeMgmtLib Include="..\sdk\$(ServiceDirectory)\*.Management.*\**\*.csproj;..\sdk\*mgmt*\**\*.csproj" />
-    <ProjectReference Include="..\sdk\$(ServiceDirectory)\**\*.csproj" Exclude="@(ExcludeMgmtLib)"/>
+  <ItemGroup Condition="'$(PackageDirectory)' == ''">
```
Comment (weshaggard) [inline eng/service.proj]: Please remove this condition we shouldn't have any references to PackageDirectory

File: sdk/servicebus/ci.yml (hunk 7 above)
Comment (weshaggard) [reply sdk/servicebus/ci.yml]: You still have file changes in here. You removed the newline at the end of the file. You should be able to use git to just undo all changes.

**Review by weshaggard — APPROVED** [review]

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml [hunk 10]
```diff
@@ -76,12 +76,19 @@ jobs:
...
-        Windows:
+          TargetFramework: netcoreapp2.1
+        Windows_NetCoreApp:
           OSName: "Windows"
           OSVmImage: "vs2017-win2016"
+          TargetFramework: netcoreapp2.1
+        Windows_NetFramework:
+          OSName: "Windows"
+          OSVmImage: "vs2017-win2016"
+          TargetFramework: net461
```
Comment (weshaggard) [inline eng/pipelines/templates/jobs/archetype-sdk-client.yml]: OK Changing this to TargetFramework broken the build, sorry about that. The reason it broken the build is because it ended up setting TargetFramework as an environment variable and msbuild picked that up and try to build some of the source projects, because of the project references from the tests, using that as the target framework. Instead lets name this TestTargetFramework which should hopefully avoid the collision with msbuild.
